src/Temporalio/Activities/ActivityDefinition.cs

```diff
@@ -24,93 +37,232 @@ private ActivityDefinition(string name, Delegate del)
-                    if (localEnd == -1)
-                    {
-                        throw new ArgumentException(
-                            $"Cannot parse name from local function {del.Method}");
-                    }
-                    name = name.Substring(0, localEnd);
+                    result = resultTaskType.GetProperty("Result")!.GetValue(resultTask);
                 }
-                // Lambdas will have >b__ on them, but we just check for the angle bracket to
-                // disallow any similar form including local functions we missed
-                if (name.Contains("<"))
+                else
                 {
-                    throw new ArgumentException(
-                        $"{del.Method} appears to be a lambda which must have a name given on the attribute");
+                    result = ValueTuple.Create();
                 }
-                if (typeof(Task).IsAssignableFrom(del.Method.ReturnType) &&
-                    name.Length > 5 && name.EndsWith("Async"))
+            }
+            return result;
+        }
+
+        private static ActivityDefinition Create(
+            MethodInfo method, bool cache, Func<object?[], object?> invoker)
+        {
+            if (cache)
+            {
+                return CachedDefinitions.GetOrAdd(method, method => Create(method, false, invoker));
+            }
+            var attr = method.GetCustomAttribute<ActivityAttribute>(false) ??
+                throw new ArgumentException($"{method} missing Activity attribute");
+            var parms = method.GetParameters();
+            return Create(
+                NameFromAttributed(method, attr),
+                method.ReturnType,
+                parms.Select(p => p.ParameterType).ToArray(),
+                parms.Count(p => !p.HasDefaultValue),
+                parameters => invoker.Invoke(ParametersWithDefaults(parms, parameters)));
+        }
+
+        private static object?[] ParametersWithDefaults(
+            ParameterInfo[] paramInfos, object?[] parameters)
+        {
+            if (parameters.Length >= paramInfos.Length)
+            {
+                return parameters;
+            }
+            var ret = new List<object?>(parameters.Length);
+            ret.AddRange(parameters);
+            for (var i = parameters.Length; i < paramInfos.Length; i++)
+            {
+                ret.Add(paramInfos[i].DefaultValue);
+            }
+            return ret.ToArray();
+        }
+
+        private static string NameFromAttributed(MethodInfo method, ActivityAttribute attr)
+        {
+            var name = attr.Name;
+            if (name != null)
+            {
+                return name;
+            }
+            // Build name from method name
+            name = method.Name;
+            // Local functions are in the form <parent>g__name|other, so we will try to
+            // extract the name
+            var localBegin = name.IndexOf(">g__");
```

Comment:
Oh boy this is nasty. Not that there's anything to be done I guess, but, eww.

Reply:
This is where I'm relying on unit tests. And note this is only for local functions which are very rare to use in .NET.